src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/ProjectFileClassifier.cs

```diff
@@ -0,0 +1,88 @@
+        /// Project files under any of these folders are considered non-modifiable.
+        /// </summary>
+        /// <remarks>
+        /// Example value: <c>"C:\Users\myusername\.nuget\;D:\LocalNuGetCache\"</c>
+        /// </remarks>
+        public string NuGetPackageFolders
+        {
+            set
+            {
+                if (!string.Equals(_nuGetPackageFoldersString, value, StringComparisons.Paths))
+                {
+                    _nuGetPackageFoldersString = value;
+                    _nuGetPackageFolders = value.Split(Delimiter.Semicolon, StringSplitOptions.RemoveEmptyEntries);
+                }
+            }
+        }
+
+        public ProjectFileClassifier()
+        {
+            _windows = Environment.GetFolderPath(Environment.SpecialFolder.Windows);
+            _programFiles86 = Environment.GetFolderPath(Environment.SpecialFolder.ProgramFilesX86);
+
+            // In a 32-bit process, SpecialFolder.ProgramFiles returns the 32-bit path.
+            // The 64-bit path is available via an environment variable however.
+            _programFiles64 = Environment.GetEnvironmentVariable("ProgramW6432");
+
+            _vsInstallationDirectory = GetVSInstallationDirectory();
+
+            return;
+
+            static string? GetVSInstallationDirectory()
+            {
+                string? dir = Environment.GetEnvironmentVariable("VSAPPIDDIR");
+
+                // The path provided is not the installation root, but rather the location of devenv.exe.
+                // __VSSPROPID.VSSPROPID_InstallDirectory has the same value.
+                // Failing a better way to obtain the installation root, remove that suffix.
+                // Obviously this is brittle against changes to the relative path of devenv.exe, however that seems
+                // unlikely and should be easy to work around if ever needed.
+                const string DevEnvExeRelativePath = "Common7\\IDE\\";
+
+                if (dir?.EndsWith(DevEnvExeRelativePath) == true)
+                {
+                    dir = dir.Substring(0, dir.Length - DevEnvExeRelativePath.Length);
+                }
+
+                return dir;
+            }
+        }
+
+        public bool IsNonModifiable(string importPath)
+        {
+            return (_programFiles64 != null && importPath.StartsWith(_programFiles64, StringComparisons.Paths))
+                   || importPath.StartsWith(_programFiles86, StringComparisons.Paths)
+                   || importPath.StartsWith(_windows, StringComparisons.Paths)
+                   || (ProjectExtensionsPath != null && importPath.StartsWith(ProjectExtensionsPath, StringComparisons.Paths))
+                   || _nuGetPackageFolders.Any(nugetFolder => importPath.StartsWith(nugetFolder, StringComparisons.Paths))
+                   || (_vsInstallationDirectory != null && importPath.StartsWith(_vsInstallationDirectory, StringComparisons.Paths));
```

Comment:
Nit: the paths that we're checking `importPath` against are all directories, right? Should we ensure that they all end in a directory separator? Otherwise we might match files that aren't under these directories.

E.g., if `C:\foo\bar\baz` is a directory, we'll think `C:\foo\bar\bazBlah.txt` is non-modifiable even though it isn't a child of `C:\foo\bar\baz`.

Reply:
Thanks, this is a good catch. Fixed.